### PR TITLE
Feed/FeedObject models & endpoints

### PR DESCRIPTION
**Pull request by kistora**
### Description of your PR

I've implemented Feed/FeedObject models & endpoints in back side.

NB:
- PAGE_LIMIT is hard coded to 40. What's your opinion ?
- Media is associated to feedObject. As a result, nobody can update or get all medias. Speaking code language, it means that there is no controller/service for Media object.

### Pre-review TODO

- [x] PR is rebased on top of `origin/master` 
- [ ] PR provide new tests for new behaviors
- [x] PR doesn't introduce commented out code
- [x] PR has been tested locally on the main use cases
- [ ] Documentation have been updated if necessary

### CODE REVIEW

File: server/app/services/feed-object-service.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  if (!currentFeedObject) throw createUserError('UnknownFeedObject', 'This feed object does not exist');
+
+  logger.verbose('FeedObject service: updating feed object named %s %s', currentFeedObject.id, currentFeedObject.title);
+
+  const transaction = await sequelize.transaction();
+
+  try {
+    if (updatedFeedObject.medias) {
+      const oldMedias = await currentFeedObject.getMedias();
+      // eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (Embraser01) [reply server/app/services/feed-object-service.js]: After some digging, its fastest and prevent some side effects

**Comment by kistora** [comment]
First of all, we schould try to use the same vocabulary:
- FeedObject: a 'news' in the feed
- Feed: a collection of feedObject

### About categories / news
The category / feedObject  association is set through the `embedded` field. It means that several categories must be created before. We schould bootstrap the categories persistance during the server startup in order to class the data coming from facebook webhook. I will list the needed categories.

### About pagination
What if we implement two query params: `page` & `limit` ?
- `page`: The current requested page number, by default page 1 is returned.
- `limit`: The current maximum number of items per response page (maximum number of items per page is 100).

### Pinned feedObjects
I think we schould create an other endpoint for pinned feedObject like `/feed/pinned`. As a result, it won't interfere with the `/feedobjects` endpoint.
K-App users (with appropriate rights) will take care of the number of pinned feedObject => so it's not a problem!

### About your comments
- I will remove `feed-controller`
- `camelCase` in wrapper table name is not consistent in our app :)
image. We could change it in a refractor branch
- Ok for HTTPS for media URL
- Category names (plural & singular) are necessary

**Comment by Embraser01** [comment]
>The category / feedObject association is set through the embedded field. It means that several categories must be created before. We schould bootstrap the categories persistance during the server startup in order to class the data coming from facebook webhook. I will list the needed categories.

So the process to set a category to a news is done manually?

>What if we implement two query params: page & limit ?
page: The current requested page number, by default page 1 is returned.
limit: The current maximum number of items per response page (maximum number of items per page is 100).

The system of pages is hard to maintain. It's easier to just set two query params `offset` & `size` which will be used as `offset` & `limit` on sequelize. Using page and limit is ambiguous (page 1 is the first or the second? if someone increase the limit, do we change page number?).

> Pinned feedObjects

Ok to create another endpoint just to pin news and getting them.

> camelCase in wrapper table name is not consistent in our app :) We could change it in a refractor branch

Actually, sequelize will probably do it for us so it's not our problem :wink:

> Category names (plural & singular) are necessary

It's weird that sequelize doesn't find automatically the plural for `Category`,the library it uses do find category <=> categories...

**Comment by kistora** [comment]
> So the process to set a category to a news is done manually?

To a feedObject I suppose ?? ... :)
It's the same process as when you want to set kommission to a barman ! Throught `_embedded` field. It imply that the category is persit before :) It's why I discuss about implement category persitance during the bootstrap process.

> Page and limit
Ok let's do it with an offset

> Plural / singular
When I remove plural / singular configuration, the `categories` field of a feedObject become `Categories`. So I need to keep the configuration.
It's not a big deal, let's do it this way!

**Comment by kistora** [comment]
My bad, it's work without name definition !

**Comment by Embraser01** [comment]
>> So the process to set a category to a news is done manually?

>To a feedObject I suppose ?? ... :)
>It's the same process as when you want to set kommission to a barman ! Throught _embedded field. It imply that the category is persit before :) It's why I discuss about implement category persitance during the bootstrap process.

I was talking about when we receive something from FB, do we set a default category, none or we try to find the best category? 😃

File: server/app/controllers/feed-object-controller.js [hunk 2]
```diff
@@ -5,22 +5,48 @@ const { createUserError } = require('../../utils');
...
+      limit: Joi.number().integer().min(1).max(100),
+    });
+
+    const { error } = schema.validate(req.query);
+    if (error) throw createUserError('BadRequest', error.message);
+
+    // Cast to number
+    offset = +req.query.offset || 0;
+    limit = +req.query.limit || 40;
   }
```
Comment (Embraser01) [inline server/app/controllers/feed-object-controller.js]: Here, `req.query` will always be true (I think an empty object is set). You could just let Joi validate the object, and after checking, you set your default values

File: server/app/models/category.js [hunk 3]
```diff
@@ -29,21 +29,16 @@ class Category extends Model {
...
 
-
   /**
    * Set associations for the model
    * @param models
    */
   static associate(models) {
     this.belongsToMany(models.FeedObject, {
-      through: 'feedobjectcategorywrapper',
+      through: 'feedobjectcategorywrappers',
```
Comment (Embraser01) [inline server/app/models/category.js]: Still no caps in here?

File: server/app/models/feed-object.js [hunk 4]
```diff
@@ -65,8 +65,9 @@ class FeedObject extends Model {
       onDelete: 'CASCADE',
     });
     this.belongsToMany(models.Category, {
-      through: 'feedobjectcategorywrapper',
+      through: 'feedobjectcategorywrappers',
```
Comment (Embraser01) [inline server/app/models/feed-object.js]: Same

File: server/app/services/feed-object-service.js [hunk 5]
```diff
@@ -151,17 +152,12 @@ async function updateFeedObject(feedObjectId, updatedFeedObject, _embedded) {
...
-        // eslint-disable-next-line
-        await newMedia.save();
-      }
-      await currentFeedObject.setMedias(updatedFeedObject.medias, { transaction });
+      await Promise.all(oldMedias.map(m => m.destroy({ transaction })));
+      const newMediasCreatePromises = updatedFeedObject.medias.map(m => m.save({ transaction }));
+      await Promise.all(newMediasCreatePromises)
+        .then(async (newMedias) => {
+          await currentFeedObject.setMedias(newMedias, { transaction });
+        });
```
Comment (Embraser01) [inline server/app/services/feed-object-service.js]: Why so much requests? We could delete every old medias in one request, then saving every updated media at once with a bulkCreate and then setting relation with a setMedias(array)?

**Comment by kistora** [comment]
> I was talking about when we receive something from FB, do we set a default category, none or we try to find the best category? 😃

We try to find the best category. For now, I list only these two categories:
- Post
- Event

File: server/app/controllers/feed-object-controller.js (hunk 2 above)
Comment (kistora) [reply server/app/controllers/feed-object-controller.js]: Ok I don't understand

File: server/app/services/feed-object-service.js (hunk 5 above)
Comment (kistora) [reply server/app/services/feed-object-service.js]: Of course :) I've never heard of bulk in sequelize !

File: server/app/controllers/feed-object-controller.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      limit: Joi.number().integer().min(1).max(100),
+    });
+
+    const { error } = schema.validate(req.query);
+    if (error) throw createUserError('BadRequest', error.message);
+
+    // Cast to number
+    offset = +req.query.offset || 0;
+    limit = +req.query.limit || 40;
+  }
```
Comment (Embraser01) [inline server/app/controllers/feed-object-controller.js]: Little example:

```js 
    const schema = Joi.object().keys({
      offset: Joi.number().integer().min(1),
      limit: Joi.number().integer().min(1).max(100),
    });
 
    const { error } = schema.validate(req.query);
    if (error) throw createUserError('BadRequest', error.message);
 
    // Cast to number
    const offset = +req.query.offset || 0;
    const limit = +req.query.limit || 40;
  
	await ...
```

File: server/app/controllers/feed-object-controller.js [hunk 7]
```diff
@@ -12,22 +12,17 @@ const Joi = require('joi');
...
+  const { error } = schema.validate(req.query);
+  if (error) throw createUserError('BadRequest', error.message);
 
-    // Cast to number
-    offset = +req.query.offset || 0;
-    limit = +req.query.limit || 40;
-  }
+  // Cast to number
+  const offset = +req.query.offset || 0;
+  const limit = +req.query.limit || 40;
```
Comment (kistora) [inline server/app/controllers/feed-object-controller.js]: No global var ?

Comment (Embraser01) [reply server/app/controllers/feed-object-controller.js]: Not for now, it is used only here so it will be simple to change it if needed

**Pull request merged**